Test the SearchShards (transport-only) API can_match functionality against searchable snapshots

quux00 opened the pull request:
The existing SearchableSnapshotsCanMatchOnCoordinatorIntegTests were all enhanced to also search via the (transport-only) SearchShards API as well as the Search API. The SearchShards API allows doing only a can-match without the other search phases.

quux00 committed:
Add IT tests for the SearchShards (transport-only) API can_match functionality against searchable snapshots

I followed the patterns in the SearchableSnapshotsCanMatchOnCoordinatorIntegTests class
and added two additional tests that use the SearchShards can_match rather than SearchRequest.

quux00 committed:
test cleanup

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

javanna commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsCanMatchOnCoordinatorIntegTests.java:
```diff
@@ -234,6 +243,207 @@ public void testSearchableSnapshotShardsAreSkippedWithoutQueryingAnyNodeWhenThey
         }
     }
 
+    public void testSearchableSnapshotShardsAreSkippedBySearchShardsAPIWithoutQueryingAnyNodeWhenTheyAreOutsideOfTheQueryRange()
```
can we add javadocs that explains what the test is doing. I can't possibly parse this super long name :) (that said, good idea to try to summarize what the test does in the method name)

javanna reviewed (approved):
LGTM thanks!

quux00 committed:
Simplified the SearchShardsAPI tests and hopefully made them more stable

quux00 commented:
@elasticmachine run elasticsearch-ci/part-2

quux00 committed:
Readjust plans again - now I embed the SearchShards API test in the existing tests

and pattern the asserts as close as possible to the Search API asserts.
In doing this I found that the only part that seems to continue to be flaky
are the expectThrows - occassionally these do not throw, but return results,
so I allow it to either throw or if can get results in those cases none
of the shards should be marked as skipped since the snapshot should be
(mostly) unavailable.
